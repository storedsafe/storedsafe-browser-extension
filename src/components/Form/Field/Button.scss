@import '../../../variables';

.field-button .field-control {
  border: 2px solid $primary;
  border-radius: $border-radius;
  background-color: $primary;
  color: $bg;
  font-weight: 800;
  text-transform: uppercase;

  &:hover,
  &:focus,
  &:active {
    color: $accent;
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  }

  &:hover {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  &:focus {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  &:active {
    border-color: $border-active;
    background-color: $bg-active;
  }
}
