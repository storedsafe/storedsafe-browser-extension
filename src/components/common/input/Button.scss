@import '../../../styles/variables.scss';

.button {
  border: 0;
  outline: 0;
  background: none;
}

.button {
  $spinner-size: $font-size;

  position: relative;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  transition: $transition-duration;
  transition-property: border-color, background-color, color, width;
  font-family: 'Roboto', sans-serif;
  padding: $spacing;
  border: $border-width solid $color-primary-0;
  background-color: $color-primary-0;
  color: $color-fg-alt;
  cursor: pointer;
  font-weight: bold;
  font-size: $font-size;
  transition: padding-right $transition-duration;

  &.button-loading {
    padding-right: $spacing * 2 + $spinner-size;
  }

  .button-spinner {
    position: absolute;
    right: $spacing;
    margin-left: $spacing;
    width: $spinner-size;
    height: $spinner-size;
    opacity: 0;
    transition: opacity $transition-duration;

    &::after {
      position: absolute;
      content: '';
      border-radius: 100%;
      box-sizing: border-box;
      left: 0;
      top: 0;
      height: 100%;
      width: 100%;
      border: $spinner-size / 6 solid #fff;
      border-top-color: $color-accent-0;
      animation: spin 1s infinite;
      transition: all $transition-duration * 3;
    }
  }

  &.button-loading .button-spinner {
    opacity: 1;
  }

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }

    to {
      transform: rotate(360deg);
    }
  }

  &:disabled {
    cursor: arrow;
  }

  &:hover:not(:disabled) {
    background-color: $color-bg;
    color: $color-primary-0;
  }

  &:focus:not(:disabled) {
    background-color: $color-bg;
    color: $color-primary-0;
  }

  &:active:not(:disabled) {
    background-color: $color-primary-1;
  }

  &.button-accent {
    background-color: $color-accent-0;
    border-color: $color-accent-0;

    &:hover {
      color: $color-accent-0;
    }

    &:focus {
      color: $color-accent-0;
    }

    &:active {
      background-color: $color-accent-1;
    }
  }

  &.button-warning {
    background-color: $color-warning-0;
    border-color: $color-warning-0;

    &:hover {
      color: $color-warning-0;
    }

    &:focus {
      color: $color-warning-0;
    }

    &:active {
      background-color: $color-warning-1;
    }
  }

  &.button-danger {
    background-color: $color-danger-0;
    border-color: $color-danger-0;

    &:hover {
      color: $color-danger-0;
    }

    &:focus {
      color: $color-danger-0;
    }

    &:active {
      background-color: $color-danger-1;
    }
  }
}
