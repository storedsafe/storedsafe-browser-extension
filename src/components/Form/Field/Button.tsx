import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Button.scss';

type ButtonProps = Pick<FieldProps, 'name' | 'type' | 'value' | 'onBlur' | 'label'>;
export const Button: React.SFC<ButtonProps | FieldAttributes> = ({
  name,
  type,
  label,
  value,
  onBlur,
  ...attributes
}: ButtonProps) => (
  <div className="field field-button">
    <button
      className="field-container field-control"
      id={name}
      name={name}
      defaultValue={value as string}
      type={type as 'button' | 'submit' | 'reset'}
      onBlur={(): void => {
        onBlur && onBlur(value, name);
      }}
      {...attributes}>
      {label}
    </button>
  </div>
);
