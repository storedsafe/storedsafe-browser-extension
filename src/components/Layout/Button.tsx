import React from 'react';
import './Button.scss';

export interface CustomButtonProps {
  color?: 'primary' | 'accent' | 'warning' | 'danger';
  isLoading?: boolean;
  children?: React.ReactNode;
  className?: string;
}

export type ButtonProps = CustomButtonProps & Omit<React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>, 'className'>;

export const Button: React.FunctionComponent<ButtonProps> = ({
  color,
  isLoading,
  children,
  className,
  ...props
}: ButtonProps) => {
  const classNames = `button button-${color}${isLoading ? ' button-loading' : ''}${className === '' ? '' : ` ${className} `}`;

  return (
    <button className={classNames} {...props}>
      {children}
      <div className="button-spinner" />
    </button>
  );
};

Button.defaultProps = {
  color: 'primary',
  isLoading: false,
  children: null,
  className: '',
};
