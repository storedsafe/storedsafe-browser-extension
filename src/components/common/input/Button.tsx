import React from 'react'
import './Button.scss'

export type ButtonColor = 'primary' | 'accent' | 'warning' | 'danger'

export interface CustomButtonProps {
  color?: ButtonColor
  isLoading?: boolean
}

type ButtonAttrs = React.ButtonHTMLAttributes<HTMLButtonElement>
type DetailedButtonAttrs = React.DetailedHTMLProps<ButtonAttrs, HTMLButtonElement>
export type ButtonProps = CustomButtonProps & DetailedButtonAttrs

export const Button: React.FunctionComponent<ButtonProps> = ({
  color,
  isLoading,
  children,
  className,
  ...props
}: ButtonProps) => {
  const classNames = [
    `button button-${color}`,
    isLoading ? 'button-loading' : '',
    className
  ].join(' ')

  return (
    <button className={classNames} {...props}>
      <div className='button-children'>{children}</div>
      <div className='button-spinner' />
    </button>
  )
}

Button.defaultProps = {
  color: 'primary',
  isLoading: false,
  children: null,
  className: ''
}
