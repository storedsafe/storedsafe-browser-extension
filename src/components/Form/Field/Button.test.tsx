import * as React from 'react';
import { Button } from './Button';
import { shallow, mount } from 'enzyme';

const onBlur = jest.fn();

beforeEach(() => {
  onBlur.mockClear();
});

test('<Button submit />', () => {
  const wrapper = shallow(<Button name="test" type="submit" label="Submit" />);
  const input = wrapper.find('button#test');
  expect(input.props().type).toBe('submit');
  expect(input.text()).toBe('Submit');
  expect(wrapper).toMatchSnapshot();
});

test('<Button button />', () => {
  const wrapper = shallow(<Button name="test" type="button" label="Button" />);
  const input = wrapper.find('button#test');
  expect(input.props().type).toBe('button');
  expect(input.text()).toBe('Button');
  expect(wrapper).toMatchSnapshot();
});

test('<Button reset />', () => {
  const wrapper = shallow(<Button name="test" type="reset" label="Reset" />);
  const input = wrapper.find('button#test');
  expect(input.props().type).toBe('reset');
  expect(input.text()).toBe('Reset');
  expect(wrapper).toMatchSnapshot();
});

test('<Button />, blur', () => {
  const wrapper = mount(<Button name="test" label="Submit" value="submit" onBlur={onBlur} />);
  wrapper.find('button#test').simulate('blur');
  expect(onBlur).toHaveBeenCalledWith('submit', 'test');
});
