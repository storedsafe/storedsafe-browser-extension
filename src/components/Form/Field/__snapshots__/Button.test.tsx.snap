// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button button /> 1`] = `ShallowWrapper {}`;

exports[`<Button reset /> 1`] = `ShallowWrapper {}`;

exports[`<Button submit /> 1`] = `ShallowWrapper {}`;
